### PR TITLE
feat: add Tag component

### PR DESCRIPTION
**Pull request by tinaClin**
# Tag

`Tag` is used for items that need to be labeled, categorized, or organized using keywords that describe them.

image
image
image

## Import

```js
import Tag from '@trendmicro/react-styled-ui/Tag';
// or
import { Tag } from '@trendmicro/react-styled-ui';
```

## Usage

```jsx
<Tag>Sample Tag</Tag>
```

### Variants

Use the `variant` prop to change the visual style of the `Tag`. You can set the value to `solid`, `outline`.

```jsx
<Stack direction="row" spacing="2x" shouldWrapChildren>
  <Tag variant="solid">Solid Tag</Tag>
  <Tag variant="outline">Outline Tag</Tag>
</Stack>
```

### Colors

Use the `variantColor` prop to change the color scheme of the `Tag`. `variantColor` can be any color key that exist in the `theme.colors`.

```jsx
<Stack spacing="4x">
  <Stack direction="row" spacing="2x" shouldWrapChildren>
    <Tag variantColor="gray">Gray</Tag>
    <Tag variantColor="red">Red</Tag>
    <Tag variantColor="magenta">Magenta</Tag>
    <Tag variantColor="purple">Purple</Tag>
    <Tag variantColor="blue">Blue</Tag>
    <Tag variantColor="green">Green</Tag>
    <Tag variantColor="teal">Teal</Tag>
    <Tag variantColor="cyan">Cyan</Tag>
  </Stack>
  <Stack direction="row" spacing="2x" shouldWrapChildren>
    <Tag variant="outline" variantColor="gray">Gray</Tag>
    <Tag variant="outline" variantColor="red">Red</Tag>
    <Tag variant="outline" variantColor="magenta">Magenta</Tag>
    <Tag variant="outline" variantColor="purple">Purple</Tag>
    <Tag variant="outline" variantColor="blue">Blue</Tag>
    <Tag variant="outline" variantColor="green">Green</Tag>
    <Tag variant="outline" variantColor="teal">Teal</Tag>
    <Tag variant="outline" variantColor="cyan">Cyan</Tag>
  </Stack>
</Stack>
```

### Round tags

```jsx
<Stack spacing="4x">
  <Stack direction="row" spacing="2x" shouldWrapChildren>
    <Tag variantColor="gray" borderRadius="lg">Gray</Tag>
    <Tag variantColor="red" borderRadius="lg">Red</Tag>
    <Tag variantColor="magenta" borderRadius="lg">Magenta</Tag>
    <Tag variantColor="purple" borderRadius="lg">Purple</Tag>
    <Tag variantColor="blue" borderRadius="lg">Blue</Tag>
    <Tag variantColor="green" borderRadius="lg">Green</Tag>
    <Tag variantColor="teal" borderRadius="lg">Teal</Tag>
    <Tag variantColor="cyan" borderRadius="lg">Cyan</Tag>
  </Stack>
  <Stack direction="row" spacing="2x" shouldWrapChildren>
    <Tag variant="outline" variantColor="gray" borderRadius="lg">Gray</Tag>
    <Tag variant="outline" variantColor="red" borderRadius="lg">Red</Tag>
    <Tag variant="outline" variantColor="magenta" borderRadius="lg">Magenta</Tag>
    <Tag variant="outline" variantColor="purple" borderRadius="lg">Purple</Tag>
    <Tag variant="outline" variantColor="blue" borderRadius="lg">Blue</Tag>
    <Tag variant="outline" variantColor="green" borderRadius="lg">Green</Tag>
    <Tag variant="outline" variantColor="teal" borderRadius="lg">Teal</Tag>
    <Tag variant="outline" variantColor="cyan" borderRadius="lg">Cyan</Tag>
  </Stack>
</Stack>
```

### Sizes

Use the `size` prop to change the size of the `Tag`. You can set the value to `sm`, `md`, or `lg`.

```jsx
<Stack spacing="4x">
  <Stack direction="row" spacing="2x" lineHeight={1} shouldWrapChildren>
    {["sm", "md", "lg"].map(size => (
      <Tag
        size={size}
        key={size}
        variantColor="gray"
      >
        { size }
      </Tag>
    ))}
  </Stack>
  <Stack direction="row" spacing="2x" lineHeight={1} shouldWrapChildren>
    {["sm", "md", "lg"].map(size => (
      <Tag
        size={size}
        key={size}
        variant="outline"
        variantColor="gray"
      >
        { size }
      </Tag>
    ))}
  </Stack>
</Stack>
```

### isCloseable

```jsx
<Stack spacing="4x">
  <Stack direction="row" spacing="2x" lineHeight={1} shouldWrapChildren>
    {["sm", "md", "lg"].map(size => (
      <Tag
        size={size}
        key={size}
        isCloseable
      >
        Gray
      </Tag>
    ))}
  </Stack>
  <Stack direction="row" spacing="2x" lineHeight={1} shouldWrapChildren>
    {["sm", "md", "lg"].map(size => (
      <Tag
        size={size}
        key={size}
        variant="outline"
        isCloseable
      >
        Gray
      </Tag>
    ))}
  </Stack>
</Stack>
```

### States

```jsx
<Stack spacing="4x">
  <Stack direction="row" spacing="4x" shouldWrapChildren>
    <Tag isCloseable>Normal</Tag>
    <Tag isCloseable disabled>Disabled</Tag>
    <Tag isCloseable invalid>Invalid</Tag>
  </Stack>
  <Stack direction="row" spacing="4x" shouldWrapChildren>
    <Tag variant="outline" isCloseable>Normal</Tag>
    <Tag variant="outline" isCloseable disabled>Disabled</Tag>
    <Tag variant="outline" isCloseable invalid>Invalid</Tag>
  </Stack>
</Stack>
```

### With input

Please reference the [`ActivedTag`](./activedtag) component.

## Props

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| disabled | boolean | | If `true`, the tag will be disabled. This sets `aria-disabled=true` and you can style this state by passing the `_disabled` prop. |
| invalid | boolean | | If `true`, set the tag to invalid style. |
| isCloseable | boolean | | If `true`, a close button will appear on the right side. |
| size | string | 'md' | The size of the tag component. One of: `'sm'`, `'md'`, `'lg'` |
| variant | string | 'solid' | The variant style of the tag component. One of: `'solid'`, `'outline'` |
| variantColor | string | 'gray' | The color scheme of the tag variant. It must be a color key defined in `theme.colors`. |

### CODE REVIEW

**Comment by tinaClin** [comment]
# ActivedTag

image

## Import

```js
import ActivedTag from '@trendmicro/react-styled-ui/ActivedTag';
// or
import { ActivedTag } from '@trendmicro/react-styled-ui';
```

## Usage

```jsx
<ActivedTag>Sample Tag</ActivedTag>
```

### With input

Set `invalid` style when value is empty after editing tag.

```jsx noInline
const EditableTag = ({
  children,
  onChange,
  ...rest
}) => {
    const [inputValue, setInputValue] = React.useState(children);
    const [inputVisible, setInputVisible] = React.useState(false);
    const handleInputVisible = () => setInputVisible(true);
    const handleInputHidden = () => setInputVisible(false);
    const handleInputBlur = (e) => {
      onChange(e);
      handleInputHidden();
    };
    const handleInputChange = (e) => {
      setInputValue(e.target.value);
    };
    const handleInputKeyUp = (e) => {
      const keyCode = event.keyCode;
      if (keyCode === 13) {
        onChange(e);
        handleInputHidden();
      }
    };
    const handleTagClick = (e) => {
      e.stopPropagation();
      handleInputVisible();
    };
    const handleTagKeyUp = (e) => {
      const keyCode = event.keyCode;
      if (keyCode === 13) {
        handleInputVisible();
      }
    };

    if (inputVisible) {
      return (
        <InputBase
          autoFocus
          minHeight="6x"
          mr="2x"
          mt="1x"
          value={inputValue}
          onChange={handleInputChange}
          onKeyUp={handleInputKeyUp}
          onBlur={handleInputBlur}
        />
      );
    }
    return (
      <ActivedTag
        mr="2x"
        mt="1x"
        onClick={handleTagClick}
        onKeyUp={handleTagKeyUp}
        {...rest}
      >
        {children}
      </ActivedTag>
    );
};

function Tags() {
  const [inputValue, setInputValue] = React.useState();
  const [inputVisible, setInputVisible] = React.useState(false);
  const [tags, setTags] = React.useState([]);
  const handleInputVisible = () => setInputVisible(true);
  const handleInputHidden = () => setInputVisible(false);
  const handleInputBlur = (e) => {
    e.stopPropagation();
    const inputValue = e.target.value;
    let newTags = tags;
    if (inputValue && tags.indexOf(inputValue) === -1) {
      newTags = [...tags, inputValue];
    }
    setTags(newTags);
    handleInputHidden();
  };
  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };
  const handleInputKeyUp = (e) => {
    const keyCode = event.keyCode;
    if (keyCode === 13) {
      const inputValue = e.target.value;
      let newTags = tags;
      if (inputValue && tags.indexOf(inputValue) === -1) {
        newTags = [...tags, inputValue];
      }
      setTags(newTags);
      handleInputHidden();
    }
  };
  const handleTagChange = (index) => (e) => {
    const inputValue = e.target.value;
    let newTags = tags;
    newTags[index] = inputValue;
    setTags([...newTags]);
  };
  const handleTagClose = (index) => (e) => {
    e.stopPropagation();
    tags.splice(index, 1);
    setTags([...tags]);
  };

  return (
    <Box
      width={320}
      minHeight={34}
      border={1}
      px="2x"
      pb="1x"
      borderColor="gray:60"
      onClick={handleInputVisible}
    >
      {
        tags.map((tag, i) => {
          return (
            <EditableTag
              key={i}
              invalid={tag.length === 0}
              onChange={handleTagChange(i)}
              onClose={handleTagClose(i)}
            >
              {tag}
            </EditableTag>
          );
        })
      }
      {(tags.length === 0 && !inputVisible) && (
        <Box
          display="inline-flex"
          alignItems="center"
          borderRadius="sm"
          border="1px dotted"
          borderColor="gray:60"
          width={92}
          Height="6x"
          py={2}
          pl="2x"
          pr="1x"
          mt="1x"
        >
          <TMIcon name="add" color="white:tertiary" />
          <Space width="2x" />
          <Text
            color="white:secondary"
            fontSize="xs"
            lineHeight="xs"
          >
            New Tag
          </Text>
        </Box>
      )}
      { inputVisible && (
        <InputBase
          autoFocus
          minHeight="6x"
          mt="1x"
          onKeyUp={handleInputKeyUp}
          onBlur={handleInputBlur}
        />
      )}
    </Box>
  );
}

render(<Tags />);
```

## Props

`ActivedTag` composes the [`Tag`](./tag) component.

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| disabled | boolean | | If `true`, the tag will be disabled. This sets `aria-disabled=true` and you can style this state by passing the `_disabled` prop. |
| invalid | boolean | | If `true`, set the tag to invalid style. |
| isCloseable | boolean | true | If `true`, a close button will appear on the right side. |
| size | string | 'md' | The size of the tag component. One of: `'sm'`, `'md'`, `'lg'` |
| variant | string | 'solid' | The variant style of the tag component. One of: `'solid'`, `'outline'` |
| variantColor | string | 'gray' | The color scheme of the tag variant. It must be a color key defined in `theme.colors`. |

File: packages/styled-ui-docs/pages/activedtag.mdx [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+## Props
+
+`ActivedTag` composes the [`Tag`](./tag) component.
+
+| Name | Type | Default | Description |
+| :--- | :--- | :------ | :---------- |
+| disabled | boolean | | If `true`, the tag will be disabled. This sets `aria-disabled=true` and you can style this state by passing the `_disabled` prop. |
+| invalid | boolean | | If `true`, set the tag to invalid style. |
+| isCloseable | boolean | true | If `true`, a close button will appear on the right side. |
```
Comment (cheton) [inline packages/styled-ui-docs/pages/activedtag.mdx]: 少了 `onClose`?

File: packages/styled-ui-docs/pages/activedtag.mdx [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+```
+
+## Props
+
+`ActivedTag` composes the [`Tag`](./tag) component.
+
+| Name | Type | Default | Description |
+| :--- | :--- | :------ | :---------- |
+| disabled | boolean | | If `true`, the tag will be disabled. This sets `aria-disabled=true` and you can style this state by passing the `_disabled` prop. |
+| invalid | boolean | | If `true`, set the tag to invalid style. |
```
Comment (cheton) [inline packages/styled-ui-docs/pages/activedtag.mdx]: 這個討論一下要換成什麼, 目前在 Input /Textarea components 是用 `aria-invalid`

File: packages/styled-ui-docs/pages/activedtag.mdx (hunk 1 above)
Comment (cheton) [reply packages/styled-ui-docs/pages/activedtag.mdx]: isCloseable 的預設值應該是 false ?

Comment (cheton) [reply packages/styled-ui-docs/pages/activedtag.mdx]: 另外我也有考慮把 Alert / AlertToast / Toast 的預設值都改成 false

Comment (tinaClin) [reply packages/styled-ui-docs/pages/activedtag.mdx]: `isCloseable `在`Tag`的預設值是false，但在`ActivedTag`裡是true喔！因為`ActivedTag`我預期是放在偽Input裡面用的，應該都會有close button吧！
image

File: packages/react-styled-ui/src/ActivedTag/index.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+const ActivedTag = forwardRef(
+  (
+    props,
+    ref,
+  ) => {
+    const {
+      isCloseable = true,
+      variantColor
+    } = props;
```
Comment (roth1002) [inline packages/react-styled-ui/src/ActivedTag/index.js]: 這邊為啥要特地跟其他地方不一樣？XDDD

```
const ActivedTag = forwardRef(
  (
    {
      isCloseable = true,
      variantColor,
      ...restProps
    },
    ref,
  ) => {
```

File: packages/react-styled-ui/src/Tag/index.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        />
+        {/* The z-index is for placing text over the above box. */}
+        <Box zIndex="1">{ children }</Box>
+        {!!isCloseable && (
+          <TagCloseButton
+            size={size}
+            disabled={disabled}
+            onClick={onClose}
+          >
+            <Icon name="_core.close-s" />
```
Comment (roth1002) [inline packages/react-styled-ui/src/Tag/index.js:67]: 這個 icon不給換囉？

File: packages/react-styled-ui/src/Tag/styles.js [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+};
+
+////////////////////////////////////////////////////////////
+
+const baseProps = {
+  alignItems: 'center',
+  verticalAlign: 'top',
+  border: 1,
+  borderColor: 'transparent',
+  cursor: 'default',
```
Comment (roth1002) [inline packages/react-styled-ui/src/Tag/styles.js:204]: 我們Box的cursor default是無效果

但Tag內的文字應該要 cursor: text吧？

File: packages/styled-ui-docs/pages/activedtag.mdx (hunk 1 above)
Comment (roth1002) [reply packages/styled-ui-docs/pages/activedtag.mdx]: 這 Focus看起來好醜...........

File: packages/styled-ui-docs/pages/tag.mdx [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+```jsx
+<Stack direction="row" spacing="2x" shouldWrapChildren>
+  <Tag variant="solid">Solid Tag</Tag>
+  <Tag variant="outline">Outline Tag</Tag>
+</Stack>
+```
+
+### Colors
+
+Use the `variantColor` prop to change the color scheme of the `Tag`. `variantColor` can be any color key that exist in the `theme.colors`.
```
Comment (roth1002) [inline packages/styled-ui-docs/pages/tag.mdx:32]: solid跟outline的VariantColor效果完全不一樣

如果沒有要分開表示的話  起碼要把你在Outline的色階告訴使用者？

螢幕快照 2020-05-25 上午10 26 21

File: packages/react-styled-ui/src/ActivedTag/index.js (hunk 3 above)
Comment (tinaClin) [reply packages/react-styled-ui/src/ActivedTag/index.js]: `isCloseable = true`嗎？因為`ActivedTag`我預期是放在偽Input裡面用的，應該都會有close button....

File: packages/react-styled-ui/src/Tag/index.js (hunk 4 above)
Comment (tinaClin) [reply packages/react-styled-ui/src/Tag/index.js:67]: 內建的close button就固定了，想要其他樣式的close button就自己放進`Tag`裡

File: packages/react-styled-ui/src/Tag/styles.js (hunk 5 above)
Comment (tinaClin) [reply packages/react-styled-ui/src/Tag/styles.js:204]: 這可以再問一下HIE，目前在zeplin上看到的是箭頭。
image

File: packages/styled-ui-docs/pages/tag.mdx (hunk 6 above)
Comment (tinaClin) [reply packages/styled-ui-docs/pages/tag.mdx:32]: 色階列出來，如果以後HIE要變更色階的話，還要記得改docs。我覺得這很容易忘記耶！

File: packages/styled-ui-docs/pages/activedtag.mdx (hunk 1 above)
Comment (tinaClin) [reply packages/styled-ui-docs/pages/activedtag.mdx]: Remove `ActivedTag` component, and add editable tag into `Tag` docs. [a3ad0f7]

File: packages/styled-ui-docs/pages/tag.mdx [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+  </Stack>
+  <Stack direction="row" spacing="4x" shouldWrapChildren>
+    <Tag variant="outline" isCloseable>Normal</Tag>
+    <Tag variant="outline" isCloseable disabled>Disabled</Tag>
+    <Tag variant="outline" isCloseable invalid>Invalid</Tag>
+  </Stack>
+</Stack>
+```
+
+### With input
```
Comment (cheton) [inline packages/styled-ui-docs/pages/tag.mdx]: 還是這段的 title 就叫做 Editable tags

File: packages/styled-ui-docs/pages/tag.mdx [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+render(<Tags />);
+```
+
+## Props
+
+| Name | Type | Default | Description |
+| :--- | :--- | :------ | :---------- |
+| disabled | boolean | | If `true`, the tag will be disabled. This sets `aria-disabled=true` and you can style this state by passing the `_disabled` prop. |
+| invalid | boolean | | If `true`, set the tag to invalid style. |
```
Comment (cheton) [inline packages/styled-ui-docs/pages/tag.mdx]: 我投 `isInvalid` 一票

File: packages/react-styled-ui/src/Tag/index.js (hunk 4 above)
Comment (cheton) [reply packages/react-styled-ui/src/Tag/index.js:67]: @tinaClin 我們要不要統一把所有的 `isCloseable` 改名成 `isCloseButtonVisible`，因為 `onClose` 其實不受這個限制，這樣意義也會更明確

File: packages/styled-ui-docs/pages/activedtag.mdx (hunk 2 above)
Comment (cheton) [reply packages/styled-ui-docs/pages/activedtag.mdx]: 交給妳另發 PR 通通改成 `isInvalid` 囉 :D

File: packages/react-styled-ui/src/Tag/index.js (hunk 4 above)
Comment (tinaClin) [reply packages/react-styled-ui/src/Tag/index.js:67]: 好的，會和`isInvalid` 發另外一個PR。

File: packages/react-styled-ui/src/ActivedTag/index.js (hunk 3 above)
Comment (tinaClin) [reply packages/react-styled-ui/src/ActivedTag/index.js]: 把`ActivedTag` component刪掉，改成example裡實作。 [a3ad0f7]

File: packages/react-styled-ui/src/Tag/styles.js (hunk 5 above)
Comment (tinaClin) [reply packages/react-styled-ui/src/Tag/styles.js:204]: 和HIE討論後，維持箭頭符號。

File: packages/styled-ui-docs/pages/tag.mdx (hunk 7 above)
Comment (tinaClin) [reply packages/styled-ui-docs/pages/tag.mdx]: Fixed in [1dda048]

File: packages/styled-ui-docs/pages/tag.mdx (hunk 8 above)
Comment (tinaClin) [reply packages/styled-ui-docs/pages/tag.mdx]: Fixed in [1dda048]

**Comment by cheton** [comment]
那個 Editable tags 妳可以留原本 Input border 的例子，然後裡面就不需要放虛線了，不然兩種範例都提供也可以，如果沒時間改可以另開別的 PR

**Review by cheton — DISMISSED** [review]

**Review by austinhung81 — DISMISSED** [review]

**Pull request merged**